src/audio/smart_amp/smart_amp_maxim_dsm.c

```diff
@@ -60,6 +60,162 @@ static int maxim_dsm_init(struct smart_amp_mod_struct_t *hspk, struct comp_dev *
+		db[(x + num_param) * DSM_PARAM_MAX + DSM_PARAM_ID] = 0x02000000 | x;
+		db[(x + num_param) * DSM_PARAM_MAX + DSM_PARAM_VALUE] = cmdblock[2];
+	}
+
+	return 0;
+}
+
+static int maxim_dsm_get_volatile_param(struct smart_amp_mod_struct_t *hspk,
+					struct comp_dev *dev)
+{
+	struct smart_amp_caldata *caldata = &hspk->param.caldata;
+	int32_t *db = (int32_t *)caldata->data;
+	enum DSM_API_MESSAGE retcode;
+	int cmdblock[1 + DSM_DEFAULT_NUM_CHANNEL];
+	int num_param = hspk->param.max_param;
+	int x;
+
+	/* Update all volatile parameter values */
+	for (x = DSM_API_ADAPTIVE_PARAM_START ; x < DSM_API_ADAPTIVE_PARAM_END ;  x++) {
+		cmdblock[0] = DSM_SET_CMD_ID(x);
+		retcode = dsm_api_get_params(hspk->dsmhandle, 1, (void *)cmdblock);
+		if (retcode != DSM_API_OK)
+			return -EINVAL;
+
+		db[x * DSM_PARAM_MAX + DSM_PARAM_ID] = 0x01000000 | x;
+		db[x * DSM_PARAM_MAX + DSM_PARAM_VALUE] = cmdblock[1];
+		db[(x + num_param) * DSM_PARAM_MAX + DSM_PARAM_ID] = 0x02000000 | x;
+		db[(x + num_param) * DSM_PARAM_MAX + DSM_PARAM_VALUE] = cmdblock[2];
+	}
+
+	return 0;
+}
+
+int maxim_dsm_get_param(struct smart_amp_mod_struct_t *hspk,
+			struct comp_dev *dev,
+			struct sof_ipc_ctrl_data *cdata, int size)
+{
+	struct smart_amp_caldata *caldata = &hspk->param.caldata;
+	size_t bs;
+	int ret = 0;
+
+	if (caldata->data) {
+		/* reset data_pos variable in case of copying first element */
+		if (!cdata->msg_index) {
+			caldata->data_pos = 0;
+			/* update volatile parameters */
+			ret = maxim_dsm_get_volatile_param(hspk, dev);
+			if (ret)
+				return -EINVAL;
+		}
+
+		bs = cdata->num_elems;
+
+		/* return an error in case of mismatch between num_elems and
+		 * required size
+		 */
+		if (bs > size) {
+			comp_err(dev, "[DSM] maxim_dsm_get_param(): invalid size %d", bs);
+			return -EINVAL;
+		}
+
+		/* copy required size of data */
+		ret = memcpy_s(cdata->data->data, size,
+			       (char *)caldata->data + caldata->data_pos,
+			       bs);
+
+		assert(!ret);
+
+		cdata->data->abi = SOF_ABI_VERSION;
+		cdata->data->size = caldata->data_size;
+		caldata->data_pos += bs;
+	} else {
+		comp_warn(dev, "[DSM] caldata->data not allocated yet.");
+		cdata->data->abi = SOF_ABI_VERSION;
+		cdata->data->size = 0;
+	}
+	return 0;
+}
+
+int maxim_dsm_set_param(struct smart_amp_mod_struct_t *hspk,
+			struct comp_dev *dev,
+			struct sof_ipc_ctrl_data *cdata)
+{
+	struct smart_amp_param_struct_t *param = &hspk->param;
+	struct smart_amp_caldata *caldata = &hspk->param.caldata;
+	/* Model database */
+	int32_t *db = (int32_t *)caldata->data;
+	/* Payload buffer */
+	int *wparam = (int *)cdata->data->data;
+	/* number of parameters to read */
+	int num_param = (cdata->num_elems >> 2);
+	int pos;
+	int x, id, ch;
+
+	if (!cdata->msg_index) {
+		pos = 0;
+		param->pos = 0;
+		param->param.id = 0;
+		param->param.value = 0;
+	} else {
+		pos = (int)(caldata->data_pos / sizeof(int32_t));
+	}
+
+	for (x = 0 ; x < num_param ; x++) {
```

Comment:
comment for the loop too. i.e. what are we doing for each param

Reply:
OK. Shall add comment here.